functions/v2/ocr/main.py

```diff
@@ -0,0 +1,212 @@
+    publish messages requesting subscribing services translate the text
+    to each target language and save the result.
+    """
+
+    print("Looking for text in image {}".format(filename))
+
+    # Use the Vision API to extract text from the image
+    image = vision.Image(
+        source=vision.ImageSource(gcs_image_uri=f"gs://{bucket}/{filename}")
+    )
+    text_detection_response = vision_client.text_detection(image=image)
+    annotations = text_detection_response.text_annotations
+
+    if len(annotations) > 0:
+        text = annotations[0].description
+    else:
+        text = ""
+    print("Extracted text {} from image ({} chars).".format(text, len(text)))
+
+    detect_language_response = translate_client.detect_language(text)
+    src_lang = detect_language_response["language"]
+    print("Detected language {} for text {}.".format(src_lang, text))
+
+    # Submit a message to the bus for each target language
+    futures = []  # Asynchronous publish request statuses
+
+    to_langs = os.environ.get("TO_LANG", "").split(",")
+    for target_lang in to_langs:
+        topic_name = os.environ.get("TRANSLATE_TOPIC")
+        if src_lang == target_lang or src_lang == "und":
+            topic_name = os.environ.get("RESULT_TOPIC")
+
+        message = {
+            "text": text,
+            "filename": filename,
+            "lang": target_lang,
+            "src_lang": src_lang,
+        }
+
+        message_data = json.dumps(message).encode("utf-8")
+        topic_path = publisher.topic_path(project_id, topic_name)
+        future = publisher.publish(topic_path, data=message_data)
+        futures.append(future)
+
+    # Wait for each publish request to be completed before exiting
+    for future in futures:
+        future.result()
+# [END functions_ocr_detect]
+
+
+# [START functions_ocr_translate]
+@functions_framework.cloud_event
+def translate_text(cloud_event):
+    """Cloud Function triggered by PubSub when a message is received from
+    a subscription.
+
+    Translates the text in the message from the specified source language
+    to the requested target language, then sends a message requesting another
+    service save the result.
+    """
+
+    # Check that the received event is of the expected type, return error if not
+    expected_type = "google.cloud.pubsub.topic.v1.messagePublished"
+    received_type = cloud_event["type"]
+    if received_type != expected_type:
+        raise ValueError("Expected {expected_type} but received {received_type}")
+
+    # Extract the message body, expected to be a JSON representation of a
+    # dictionary, and extract the fields from that dictionary.
+    data = cloud_event.data["message"]
+    try:
+        message_data = base64.b64decode(data)
+        message = json.loads(message_data)
+
+        text = message["text"]
+        filename = message["filename"]
+        target_lang = message["lang"]
+        src_lang = message["src_lang"]
+    except Exception as e:
```

Comment:
There are a few that could happen. Since it's a test, I took a broad brush and the tester can see what happened in the output. Note that no exception should occur.

Reply:
If we're expecting a bunch, SGTM to keep it as is.